src/test/java/io/pravega/connectors/flink/FlinkPravegaReaderITCase.java

```diff
@@ -78,6 +84,75 @@ public void testMultipleSourcesMultipleSegments() throws Exception {
     }
 
+    @Test
+    public void testWatermark() throws Exception {
+        // set up the stream
+        final String streamName = RandomStringUtils.randomAlphabetic(20);
+        SETUP_UTILS.createTestStream(streamName, 4);
+        try (
+                final EventStreamWriter<Integer> eventWriter = SETUP_UTILS.getIntegerWriter(streamName);
+
+                // create the producer that writes to the stream
+                final ThrottledIntegerWriter producer = new ThrottledIntegerWriter(
+                        eventWriter,
+                        NUM_STREAM_ELEMENTS,
+                        NUM_STREAM_ELEMENTS + 1,  // no need to block
+                        0,                 // the sleep time per element
+                        true
+                )
+        ) {
+            producer.start();
+            producer.sync();
+
+            final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+            env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
+            env.setParallelism(4);
+            env.getConfig().setAutoWatermarkInterval(1000);
+
+            // the Pravega reader
+            final FlinkPravegaReader<Integer> pravegaSource = FlinkPravegaReader.<Integer>builder()
+                    .forStream(streamName)
+                    .enableMetrics(false)
+                    .withPravegaConfig(SETUP_UTILS.getPravegaConfig())
+                    .withDeserializationSchema(new IntegerDeserializationSchema())
+                    .withTimestampAndWatermark(new LowerBoundAssigner<Integer>() {
+                        @Override
+                        public long extractTimestamp(Integer element, long previousElementTimestamp) {
+                            return element;
+                        }
+                    })
+                    .build();
+
+            env
+                    .addSource(pravegaSource)
+                    .timeWindowAll(Time.seconds(1))
+                    .apply(new AllWindowFunction<Integer, Integer, TimeWindow>() {
+                        @Override
+                        public void apply(TimeWindow window, Iterable<Integer> arr, Collector<Integer> collector) throws Exception {
+                            for (Integer i : arr) {
+                                collector.collect(i);
+                            }
+                        }
+                    })
+                    .addSink(new IntSequenceExactlyOnceValidator(NUM_STREAM_ELEMENTS))
+                    .setParallelism(1);
+
+            final long executeStart = System.nanoTime();
+
+            // if these calls complete without exception, then the test passes
+            try {
+                env.execute();
+            } catch (Exception e) {
+                if (!(ExceptionUtils.getRootCause(e) instanceof SuccessException)) {
+                    throw e;
+                }
+            }
+
+            final long executeEnd = System.nanoTime();
```

Comment:
Where are the assertions about the correctness ?

Reply:
The `IntSequenceExactlyOnceValidator` guarantees that no data loss or duplicate.
The `ExceptionUtils.getRootCause(e) instanceof SuccessException` guarantees that no error in the processing pipeline.

Reply:
Brian is correct, the `IntSequenceExactlyOnceValidator` is checking for correctness and either calling `Assert.fail` if there's a problem or throwing a `SuccessException` if all expected messages are received in the correct order.

That said, it's confusing if it DOES fail because the original `Assert.fail` message is lost with the reason WHY it failed.  I've now fixed this with an update that rethrows the original `AssertException` if that's the reason it failed.